### ISSUE TITLE
[feature] Add support for Tensorboard in KFP v2

### ISSUE THREAD

**Issue by AlexandreBrown** [kubeflow/pipelines#8794]
### Feature Area

/area frontend 
/area backend
/area sdk

### What feature would you like to see?
As a Kubeflow Pipelines user, I would like to be able to use the SDK to log a list of numerical values and to be able to view it in Kubeflow UI and compare multiple runs to see how they compare.  
It is very common when performing ML trainings to want to log some metrics over time eg: log a list of accuracy values over the epochs during training to get a learning curve or simply the `train_loss` and `val_loss` for instance.  
This can also be used to compare how 2 models differ (eg: accuracy over time of model 1 vs accuracy over time of model 2).    
Example:  
<img src="https://user-images.githubusercontent.com/26939775/216708839-ad8a9135-af31-43fc-8ad9-6b66b48fc36f.png" height="300"/>  
<img src="https://user-images.githubusercontent.com/26939775/216712850-9658c74b-f976-41c7-a7db-cfde32dc15a7.png" height="300"/>  

### What is the use case or pain point?

The use case is to be able to easily compare runs for metrics that are more meaningful when represented as a list of numerical values instead of a single value.    
Currently we can use `Output[Metrics]` to store a single key value pair metric (eg: using key=`train_accuracy` and value=`123`) or we can log a `dict`, the first approach does not allow to get all the values and the second approach is not suitable because it is super hard to reason about.  
Here is an example of the two approaches (with random values just to illustrate the point):    
<img src="https://user-images.githubusercontent.com/26939775/216711722-e789df50-59ed-4fcb-84c4-3df5f1281f67.png" height="100"/>  
<img src="https://user-images.githubusercontent.com/26939775/216711611-9b86e58d-edde-4b2c-87a7-51702a34d3f1.png" height="300"/>  
 
Adding this feature could allow the possibility to offer an option to overlap the 2 scatter plots of 2 runs for easier comparison (should remain an optional button because in some case it's easier to view the 2 plots separately).  

### Is there a workaround currently?

Right now the workaround is to plot the values (eg: using matplotlib) during a pipeline step then to convert the matplotlib figure to an image (eg: jpg or png) then to convert the image to a base64 string that can be used in an `<img>` HTML tag that we can then log by using `Output[Markdown]` or `Output[HTML]`.  
This is works but it has downsides :  
- Does not support to dynamically scrub the values when viewing the visualisations after the execution of the pipeline  
- Cannot be used to overlap 2+ runs and see how they compare for the same metric
- Requires some boilerplate code (plotting, converting to an image, image to base64, html etc)
  
Adding this feature would also keep Kubeflow competitive with other solutions which offer this kind of feature (eg: w&b, dvc pipelines, MLFlow etc) 

---

Love this idea? Give it a 👍.

**Comment by AlexandreBrown** [comment]
@zijianjoy Hello, I would like to hear your thoughts on this feature request, do you think it would be feasible in KFP?.  
Thanks

**Comment by zijianjoy** [comment]
I think this feature is already supported for ROC curve, confusion matrix, and scalar metrics. Can you check if the followings match your need?

Compare pipeline runs: https://www.kubeflow.org/docs/components/pipelines/v1/sdk-v2/run-comparison/

How to create V2 metrics: https://www.kubeflow.org/docs/components/pipelines/v1/sdk/output-viewer/#v2-sdk-use-sdk-visualization-apis

**Comment by AlexandreBrown** [comment]
@zijianjoy thank you for your feedback.   
To the best of my knowledge, I do not think this feature is already supported given the currently available output types.  
Let me try to explain why it is not the case for each of them :  
- A ROC Curve is meant to be used to asses the trade-off between the true positive rate and false positive rate for a model using a treshold.  
  - If we take the use case described by this feature request, if we wanted to plot the `train_loss` vs the `val_loss` metrics (which under the hood is just two list of floats), we would not be able to since there is no concept of true positive and false positive in this case.
  - The use case described by the feature request is to be able to plot a list of floats regardless of its underlying semantic, so a list of `train_loss` or `val_loss` is one example, a list of accuracy values (floats) is another example.  
  - We could have KFP set the X axis to numeric values from 1 to N where N is the number of elements in the list we are asking to plot or the sdk could allow to set the custom X axis values if the default does not suit us eg: `metrics.log_metric(y=accuracy_values, x=numpy.arange(start=1, stop=len(accuracy_values), step=1)`  
  would be the same as : `metrics.log_metric(y=accuracy_values)` by default. 
-  A confusion matrix is meant to be used to represent the performance of a model for various categories of labels (eg: It can display if a model predicted cat 7/10 times when the true value was cat and 3/10 times it predicted cat when the true value was dog etc).
  - It is not meant to display a list of values (eg: list of floats like for `train_loss`, `val_loss`, accuracy value**s** list etc.
  - The feature request suggests an approach that can plot arbitrary list of values for any type of models. We could also extend this to support multiple list of values like matplotlib plot can. For instance if we want to plot the predicted stock price values vs the actual prices, we could do so by calling the `metrics.log_metric` method twice, once with the predicted y and another time with the actual y. That's just one example, another example could be to plot the predicted vs actual values for multiple methods (multiple predicted y), same principe, we would call the `metrics.log_metric` for the actual and for each predicted values.  
  -   
-

**Comment by AlexandreBrown** [comment]
@zijianjoy thank you for your feedback.   
To the best of my knowledge, I do not think this feature is already supported given the currently available output types.  
Let me try to explain why it is not the case for each of them :  
- A ROC Curve is meant to be used to asses the trade-off between the true positive rate and false positive rate for a model using a treshold.  
  - If we take the use case described by this feature request, if we wanted to plot the `train_loss` vs the `val_loss` metrics (which under the hood is just two list of floats), we would not be able to since there is no concept of true positive and false positive in this case.
![image](https://user-images.githubusercontent.com/26939775/217280295-10c3a7c5-f405-46f9-bbe4-5494de988325.png)
  - The use case described by the feature request is to be able to plot a list of floats regardless of its underlying semantic, so a list of `train_loss` or `val_loss` is one example, a list of accuracy values (floats) is another example.
![image](https://user-images.githubusercontent.com/26939775/217280619-c17c81ae-d47d-4904-9b88-344ac204a537.png)
  - We could have KFP set the X axis to numeric values from 1 to N by default where N is the number of elements in the list we are asking to plot or the sdk could allow to set the custom X axis values if the default does not suit us eg: `metrics.log_metric(y=accuracy_values, x=numpy.arange(start=1, stop=len(accuracy_values), step=1)`  
  would be the same as : `metrics.log_metric(y=accuracy_values)` by default.  

-  A confusion matrix is meant to be used to represent the performance of a model for various categories of labels (eg: It can display if a model predicted cat 7/10 times when the true value was cat and 3/10 times it predicted cat when the true value was dog etc).
    - It is not meant to display a list of values (eg: list of floats like for `train_loss`, `val_loss`, accuracy value**s** list etc). 
    - The feature request suggests an approach that can plot arbitrary list of values for any type of models. We could also extend this to support multiple list of values like matplotlib plot can. For instance if we want to plot the predicted stock price values vs the actual prices, we could do so by calling the `metrics.log_metric` method twice, once with the predicted y and another time with the actual y. That's just one example, another example could be to plot the predicted vs actual values of various methods/models (multiple predicted y), same principe, we would call the `metrics.log_metric` for the actual values and once for each predicted values list.  
Here is an example I have already done in the past : 
![image](https://user-images.githubusercontent.com/26939775/217279403-c9d2fa8f-408f-4a62-a9f4-7b358e05f0f5.png)
- As for the scalar metrics, it is what was described in this feature request as being one approach that is currently possible but does not hit home.
  - This approach would not plot the values but rather only show one value or the list of values but would not be presented in a plot making it impossible to reason about.  
![image](https://user-images.githubusercontent.com/26939775/217281733-cd8bd300-3601-4469-8d85-3fd9eead9a1c.png)
  - If this output metric type would plot a list of values instead of showing the raw values it would be a first step in the right direction but considering the other usages (like multiple list of values on the same plot), the KFP sdk should check if creating a new output metric type would be easier (I do not know if creating a new type or adapting this one would be easier to implement in the sdk so feel free to correct me).
   
Hopefully this clarifies the intent of the feature requested here.  
  
Let me know what you think :)

**Comment by AlexandreBrown** [comment]
CC @chensun

**Comment by zijianjoy** [comment]
Thank you for the explanation, if I understand correctly, there is a need for plot drawing metrics, which is generic enough that users can specify the `name of metrics`, `number of metrics`, `Name of x/y axis`, `upper bound and lower bound of x/y axis`, `x/y axis value for each numerical metrics`. 

Please confirm if this understanding is correct, and whether you think it is valuable to compare two or more of such numerical metrics?

**Comment by zijianjoy** [comment]
If you are interested in making contribution, will you be able to start a design doc for this feature? I feel that there can be multiple options to implement this.

**Comment by AlexandreBrown** [comment]
> if I understand correctly, there is a need for plot drawing metrics, which is generic enough that users can specify the name of metrics, number of metrics, Name of x/y axis, upper bound and lower bound of x/y axis, x/y axis value for each numerical metrics.
Please confirm if this understanding is correct  

@zijianjoy  Yes, your understanding is correct this is exactly it.   
I think that there are some interesting decisions to make as to what must be provided by the sdk user and what can be inferred and used as default values.   
For instance, here are some examples of decisions that could be discussed and analyzed to enhance the SDK user experience & Kubeflow UI/UX:  
- One might say we can infer the number of metrics from the number of logged metrics.  
- `Name of x/y axis` could be made optional and left blank if not provided.  
- `upper bound and lower bound of x/y axis` can be inferred to the min and max of their respective axis from all the values provided to ensure all values are visible by default but the user could change the default. For this one, we could also imagine a future improvement that would allow the user to change these upper/lower bounds when visualizing the UI so that outliers do not make visualization hard for instance. This would not change the underlying data, pure frontend feature.  
- `x/y axis value for each numerical metrics`, for this one, I think only the x axis could be inferred to be (1 .. N) where N is the maximum number of values present in the logged metrics but the user should be able to change the default if needed.  
  
Of course these are just examples and more thoughts should go into those details but I think it's also fair to not expect to be right the first time and that most likely these defaults can improve with time and community feedback.  

> whether you think it is valuable to compare two or more of such numerical metrics on UI?  

I think it would be extremely valuable to be able to compare at least two of such numerical metrics on the UI, for ease of comparison, we could again make the range of the x and y axis be the min and max of the x and y axis for all logged metrics.   
A clear use case of this (but use cases are not limited to this example) would be to compare the same numerical metrics for two models, this is fairly common in experiment tracking solutions :   
<img src="https://user-images.githubusercontent.com/26939775/217971306-2c2d45a1-1f97-4e46-83f2-9db3a7241ceb.png" height="300"/>  
  
> If you are interested in making contribution, will you be able to start a design doc for this feature? I feel that there can be multiple options to implement this.  
  
Yes I am willing to create such document.  
I think you are right, there are a lot of options each with their pros and cons.  
This will allow the community to weigh in on some of these options and implementation details that might affect KFP users experience.

**Comment by AlexandreBrown** [comment]
@zijianjoy Here is the link to the [design doc](bit.ly/3HSK5i9) so far.  
Let me know what you think and feel free to request access to it so I can grant you editor access.

**Comment by AlexandreBrown** [comment]
Hello @zijianjoy , would it be possible to get an update on my latest replies https://docs.google.com/document/d/1mVsV4C02stBlamhQm0bPeBsFDBsXdQXD7gyMJ4l-A6s/edit?usp=drivesdk
If my understanding of your proposal is correct, it seems like your proposal would greatly reduce the scope and therefore simplify the implementation of the feature.
What do you think? Is that right?

**Comment by zijianjoy** [comment]
Thank you @AlexandreBrown for the design doc! 

To answer your question: Yes you captured my suggestion correctly and thanks for adding this point to the design doc.

Also, I have brought this to the KFP team for discussion. 
We think that the generic numerical values metrics can be useful, but it might not be able to cover all use cases because:
- The data format for numerical values metrics is an additional workload for users to learn and adopt.
- If users need a different visualization approach, it is hard to adjust existing data format to adopt new requirements.
- We have existing functionality in KFPv1 like Tensorboard which can provide same and more advanced support: https://www.kubeflow.org/docs/components/pipelines/v1/sdk/output-viewer/#tensorboard. 

Would you like to evaluate whether Tensorboard can fulfill your need? If yes, this feature request will be about supporting Tensorboard in KFPv2. And we welcome contributions for this feature.

**Comment by AlexandreBrown** [comment]
@zijianjoy 
Being able to have the Tensorboard data for a run and see the run metrics is important and useful.
In addition to what was possible with v1, I think it would be nice if KFP v2 had a way to aggregate the Tensorboard logs of multiple runs to then offer a Tensorboard compare view with data from multiple runs.

Under the hood it could download the data from S3 of multiple runs then create a folder with all runs data. Then once all the data is in one folder it could treat it just like if ot was a single run and display a single Tensorboard dashboard that would show all the runs metrics.

**Comment by yingding** [comment]
It would be nice to have Tensorboard gaining direct access to a KFP artifact, which maybe kind of model training log files that Tensorboard can deal with, and the Tensorboard can be tight integrated into visualization section of the KFP components.  

I had also issue to use an on-prem s3 bucket with ssl certificate validation turned off in KFP v1, while writing logs to s3 bucket from KFP component is also a lot template coding. These all will be benefited from an additional `Output[TFLog]` in V2.

Please let me know your thoughts on this matter.
